README.md
# mi-programoptions

This library is a small command line argument parser library for C++11
and later:

- it only handles strings
- it supports long options with --, short options with -, and
  positional arguments
- it avoids option name strings repetition by using option objects
  such that the compiler can find some typos

The library has no runtime dependencies beyond the standard C++
library.

The library uses mi-cpptest for the optional unit tests. In the CMake
build, unit tests may be disabled by setting `ENABLE_TESTS` to `0`.

## Use with CMake

1. either include this as a subproject with `ADD_SUBDIRECTORY(...)`,
   which will build a static library
2. or build and install and use `FIND_PACKAGE(mi-programoptions)`,
   which will build a shared library

In both cases, link to `mi-programoptions`.

## Use without CMake

As the library consists of only 3 code files, 1 of which tiny, it
should be easy to use with other build systems.


debian/changelog
mi-programoptions (2.0.1-1) unstable; urgency=low

  * new upstream version

 -- Alexander Bürger <alexander.buerger@met.no>  Fri, 9 Sep 2022 08:00:08 +0200

mi-programoptions (2.0.0-1) unstable; urgency=low

  * new upstream version

 -- Alexander Bürger <alexander.buerger@met.no>  Thu, 14 Oct 2021 08:00:08 +0200

mi-programoptions (1.0.0-7) unstable; urgency=low

  * include shared library in package

 -- Alexander Bürger <alexander.buerger@met.no>  Fri, 26 Mar 2021 08:00:08 +0100

mi-programoptions (1.0.0-6) unstable; urgency=low

  * use correct soversion for dbg package when calling dh_strip

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Mar 2021 08:00:08 +0100

mi-programoptions (1.0.0-5) unstable; urgency=low

  * allow correct lib packages soversion

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Mar 2021 08:00:08 +0100

mi-programoptions (1.0.0-4) unstable; urgency=low

  * check all lib packages vs soversion

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Mar 2021 08:00:08 +0100

mi-programoptions (1.0.0-2) unstable; urgency=low

  * check lib package vs soversion

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 18 Mar 2021 08:00:08 +0100

mi-programoptions (1.0.0-1) unstable; urgency=low

  * new upstream version

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 09 Mar 2021 08:00:08 +0100

mi-programoptions (0.1.0-1) unstable; urgency=low

  * debian packaging

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 24 Apr 2019 08:00:08 +0200


test_programoptions.cc

#include "mi_cpptest.h"

#include "mi_programoptions.h"

using namespace miutil::program_options;

MI_CPPTEST_TEST_CASE(progopt_config_file)
{
    const option o1("one.setting", "this is a setting");
    const option o2("one.option", "this is an option");
    const option o3("no_dot", "this is special");

    option_set options;
    options.add(o1).add(o2).add(o3);

    std::istringstream configfile("no_dot=5\n[one]\nsetting=hei\noption=hi\n");
    const value_set values = parse_config_file(configfile, options);

    MI_CPPTEST_CHECK(values.is_set(&o1));
    MI_CPPTEST_CHECK_EQ("hi", values.value(&o2));
    MI_CPPTEST_CHECK_EQ("5", values.value(&o3));
}

MI_CPPTEST_TEST_CASE(progopt_cmdline_and_dump)
{
  const option o1 = option("one.setting", "this is a setting").set_composing();
  const option o2("one.option", "this is an option");

  option_set options;
  options.add(o1).add(o2);

  const char* argv[] = { "test.exe", "--one.setting", "hei", "--one.setting=hi", "hey" };
  const int argc = sizeof(argv) / sizeof(argv[0]);

  string_v positional;
  const value_set values = parse_command_line(argc, const_cast<char**>(argv), options, positional);

  MI_CPPTEST_CHECK_EQ("hei", values.value(&o1, 0));
  MI_CPPTEST_CHECK_EQ("hi", values.value(&o1, 1));
  MI_CPPTEST_CHECK(!values.is_set(&o2));
  MI_CPPTEST_CHECK_EQ(1, positional.size());
  MI_CPPTEST_CHECK_EQ("hey", positional[0]);

  std::ostringstream dump;
  options.dump(dump, values);
  MI_CPPTEST_CHECK_EQ("--one.setting\n  => 'hei'\n  => 'hi'\n", dump.str());
}

MI_CPPTEST_TEST_CASE(progopt_values_add)
{
  const option o1("one.setting", "this is a setting");
  const option o2("one.option", "this is an option");

  option_set options;
  options.add(o1).add(o2);

  std::istringstream configfile1("[one]\nsetting=hei\n");
  const value_set values1 = parse_config_file(configfile1, options);
  MI_CPPTEST_CHECK(values1.is_set(&o1));

  std::istringstream configfile2("[one]\noption=hi\n");
  const value_set values2 = parse_config_file(configfile2, options);
  MI_CPPTEST_CHECK_EQ("hi", values2.value(&o2));

  value_set values = values1;
  values.add(values2);
  MI_CPPTEST_CHECK_EQ("hei", values.value(&o1));
  MI_CPPTEST_CHECK_EQ("hi", values.value(&o2));
  MI_CPPTEST_CHECK_EQ(&o2, values.find("one.option", false));
  MI_CPPTEST_CHECK_EQ(values.find("no.way", false), nullptr);
}

MI_CPPTEST_TEST_CASE(progopt_help)
{
  const option o1 = option("one.setting", "this is a setting").set_shortkey("os");
  const option o2 = option("one.option", "this is an option").set_default_value("hi");

  option_set options;
  options.add(o1).add(o2);

  std::ostringstream help;
  options.help(help);
  MI_CPPTEST_CHECK_EQ("--one.setting / -os: this is a setting\n--one.option: this is an option (default: hi)\n", help.str());
}

MI_CPPTEST_TEST_CASE(progopt_short)
{
  const option o1 = option("one.setting", "this is a setting").set_composing().set_shortkey("os");
  const option o2("one.option", "this is an option");

  option_set options;
  options.add(o1).add(o2);

  const char* argv[] = { "test.exe", "-os", "hei", "--one.setting=hi" };
  const int argc = sizeof(argv) / sizeof(argv[0]);

  string_v positional;
  const value_set values = parse_command_line(argc, const_cast<char**>(argv), options, positional);

  MI_CPPTEST_CHECK_EQ("hei", values.value(&o1, 0));
  MI_CPPTEST_CHECK_EQ("hi", values.value(&o1, 1));
  MI_CPPTEST_CHECK(!values.is_set(&o2));
}

MI_CPPTEST_TEST_CASE(progopt_narg)
{
  const option o1 = option("help", "show help").set_shortkey("h").set_narg(0);
  const option o2 = option("this.option", "this expects two").set_shortkey("to").set_narg(2);

  option_set options;
  options.add(o1).add(o2);

  const char* argv[] = { "test.exe", "-to", "hei", "ho", "-h", "extra" };
  const int argc = sizeof(argv) / sizeof(argv[0]);

  string_v positional;
  const value_set values = parse_command_line(argc, const_cast<char**>(argv), options, positional);

  MI_CPPTEST_CHECK(values.is_set(&o1));
  MI_CPPTEST_CHECK_EQ("hei", values.value(&o2, 0));
  MI_CPPTEST_CHECK_EQ("ho", values.value(&o2, 1));
  MI_CPPTEST_CHECK_EQ("extra", positional.at(0));
}

MI_CPPTEST_TEST_CASE(progopt_consume_positional)
{
  const option o1 = option("one.setting", "this is a setting").set_composing();
  const option o2("one.option", "this is an option");

  option_set options;
  options.add(o1).add(o2);

  const char* argv[] = { "test.exe", "--one.setting", "hei", "--one.setting=hi", "hey" };
  const int argc = sizeof(argv) / sizeof(argv[0]);

  string_v positional;
  value_set values = parse_command_line(argc, const_cast<char**>(argv), options, positional);

  MI_CPPTEST_CHECK_EQ("hei", values.value(&o1, 0));
  MI_CPPTEST_CHECK_EQ("hi", values.value(&o1, 1));
  MI_CPPTEST_CHECK(!values.is_set(&o2));

  positional_args_consumer pac(values, positional);
  pac >> o2;
  MI_CPPTEST_CHECK(pac.done());
  MI_CPPTEST_CHECK_EQ("hey", values.value(o2));
}

MI_CPPTEST_TEST_CASE(progopt_end_of_options)
{
  const option o1 = std::move(option("one.setting", "this is a setting").set_composing());
  const option o2("one.option", "this is an option");

  option_set options;
  options.add(o1).add(o2);

  const char* argv[] = { "test.exe", "--one.setting", "hei", "--", "--one.option=hi", "hey" };
  const int argc = sizeof(argv) / sizeof(argv[0]);

  string_v positional;
  value_set values = parse_command_line(argc, const_cast<char**>(argv), options, positional);

  MI_CPPTEST_CHECK_EQ(1, values.values(&o1).size());
  MI_CPPTEST_CHECK_EQ(2, positional.size());
}

MI_CPPTEST_TEST_CASE(progopt_multiple_keys)
{
  const option o1 = option("one.setting", "this is a setting").add_key("one.choice");

  option_set options;
  options.add(o1);

  std::istringstream configfile1("[one]\nchoice=hei\n");
  const value_set values1 = parse_config_file(configfile1, options);
  MI_CPPTEST_CHECK(values1.is_set(&o1));
}

MI_CPPTEST_TEST_CASE(progopt_multiple_shortkeys)
{
  const option o2 = option("one.option", "this is an option").add_shortkey("option");

  option_set options;
  options.add(o2);

  std::vector<std::string> cmdline {"test.exe", "-option=hi"};
  string_v positional;
  const value_set values2 = parse_command_line(cmdline, options, positional);
  MI_CPPTEST_CHECK_EQ("hi", values2.value(&o2));
}

MI_CPPTEST_TEST_CASE(progopt_bad_duplicate)
{
  const option os = option("setup", "this is an option").add_shortkey("s");

  option_set options;
  options.add(os);

  std::vector<std::string> cmdline {"test.exe", "--setup=ha", "-s", "he"};
  string_v positional;
  MI_CPPTEST_CHECK_THROW(parse_command_line(cmdline, options, positional), option_error);
}

MI_CPPTEST_TEST_CASE(progopt_overwriting)
{
  const option os = std::move(option("setup", "this is an option").add_shortkey("s").set_overwriting());

  option_set options;
  options.add(os);

  std::vector<std::string> cmdline {"test.exe", "--setup=ha", "-s", "he"};
  string_v positional;
  const value_set values2 = parse_command_line(cmdline, options, positional);
  MI_CPPTEST_CHECK_EQ("he", values2.value(&os));
}


mi_po_parse.cc
/*
  mi-programoptions

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-programoptions.

  mi-programoptions is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-programoptions is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-programoptions; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_programoptions.h"

#include <fstream>

namespace {
const std::string EMPTY;
}

namespace miutil {
namespace program_options {

value_set parse_config_file(const std::string& filename, option_set& options)
{
  std::ifstream infile(filename);
  if (!infile)
    throw option_error("cannot read config file '" + filename + "'");

  try {
    return parse_config_file(infile, options);
  } catch (option_error& oe) {
    throw option_error("while reading '" + filename + ": " + oe.what());
  }
}

#define OPTION_CHARS "[a-zA-Z0-9._][a-zA-Z0-9._-]*"

value_set parse_config_file(std::istream& infile, option_set& options)
{
  value_set values;
  static const std::regex re_comment("\\s*#.*");
  static const std::regex re_section("\\[(" OPTION_CHARS ")\\]$");
  enum { GS_ALL, GS_KEY };
  static const std::regex re_value("\\s*(" OPTION_CHARS ")\\s*=\\s*(.*)?\\s*");
  enum { GV_ALL, GV_KEY, GV_VALUE };
  std::string line;
  std::string section;
  for (int lineno = 1; std::getline(infile, line); ++lineno) {
    if (line.empty())
      continue;
    std::smatch match;
    if (std::regex_match(line, match, re_comment)) {
      // ignore
    } else if (std::regex_match(line, match, re_section)) {
      section = match[GS_KEY].str() + ".";
    } else if (std::regex_match(line, match, re_value)) {
      std::string value = match[GV_VALUE].str();
      if (value.size() >= 2 && (value[0] == '\'' || value[0] == '"') && value[0] == value[value.size() - 1])
        value = value.substr(1, value.size() - 2);
      try {
        values.put(options.find_option(section + match[GV_KEY].str(), false), value);
      } catch (option_error& oe) {
        std::ostringstream msg;
        msg << "lineno " << line << ": " << oe.what();
        throw option_error(msg.str());
      }
    } else {
      std::ostringstream msg;
      msg << "bad line " << lineno << ": " << line;
      throw option_error(msg.str());
    }
  }
  if (!infile.eof() && infile.bad())
    throw option_error("error reading config");
  return values;
}

value_set parse_command_line(const std::vector<std::string>& argv, option_set& options, std::vector<std::string>& positional)
{
  value_set values;
  static const std::regex re_option("(-{1,2})(" OPTION_CHARS ")(=(.*))?");
  enum { GO_ALL, GO_DASH, GO_KEY, GO_EQ_VALUE, GO_VALUE };
  bool end_of_options_marker = false;
  const int argc = argv.size();
  for (int a = 0; a < argc; ++a) {
    const std::string arg(argv[a]);
    if (arg == "--") {
      end_of_options_marker = true;
      continue;
    }
    std::smatch match;
    if (!end_of_options_marker && std::regex_match(arg, match, re_option)) {
      option_cx opt = options.find_option(match[GO_KEY], match[GO_DASH].length() == 1);
      if (match[GO_VALUE].matched) {
        if (opt->narg() == 0) {
          throw option_error("arg for no-arg option '" + opt->key() + "'");
        } else if (!opt->is_composing() && opt->narg() != 1) {
          std::ostringstream msg;
          msg << "args for option '" + opt->key() + "', which expects exactly " << opt->narg() << " values, cannot be passed with '='";
          throw option_error(msg.str());
        }
        values.put(opt, match[GO_VALUE].str());
      } else if (opt->has_implicit_value()) {
        values.put_implicit(opt);
      } else if (opt->narg() == 0) {
        values.put(opt, std::string());
      } else if (opt->is_composing() && a + 1 < argc) {
        values.put(opt, argv[++a]);
      } else if (!opt->is_composing() && a + (int)opt->narg() < argc) {
        string_v args;
        args.reserve(opt->narg());
        const int alast = a + opt->narg();
        while (a < alast)
          args.push_back(argv[++a]);
        values.put(opt, args);
      } else {
        throw option_error("no arg for option '" + opt->key() + "'");
      }
    } else {
      positional.push_back(arg);
    }
  }
  return values;
}

value_set parse_command_line(int argc, char* argv[], option_set& options, std::vector<std::string>& positional)
{
  return parse_command_line(std::vector<std::string>(argv + 1, argv + argc), options, positional);
}

positional_args_consumer& positional_args_consumer::operator>>(const option& opt)
{
  if (opt.is_composing()) {
    values_.put(&opt, *pbegin_++);
  } else if (!values_.is_set(opt)) {
    if (std::distance(begin(), end()) >= (int)opt.narg()) {
      const string_v::const_iterator opt_end = begin() + opt.narg();
      values_.put(&opt, string_v(begin(), opt_end));
      pbegin_ = opt_end;
    } else {
      throw option_error("positional arg error for option '" + opt.key() + "'");
    }
  }
  return *this;
}

void positional_args_consumer::dump(std::ostream& out) const
{
  out << "remaining positional arguments: " << std::endl;
  for (string_v::const_iterator p = pbegin_; p != positional_.end(); ++p)
    out << " -> '" << *p << "'" << std::endl;
}

} // namespace program_options
} // namespace miutil


mi_po_option.cc
/*
  mi-programoptions

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-programoptions.

  mi-programoptions is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-programoptions is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-programoptions; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_programoptions.h"

namespace {
const std::string EMPTY;
}

namespace miutil {
namespace program_options {

struct option::d
{
  std::vector<std::string> keys_;
  std::vector<std::string> shortkeys_;
  std::string help_;
  size_t narg_;
  bool is_composing_;
  bool is_overwriting_;
  std::string default_;
  bool has_default_;
  std::string implicit_;
  bool has_implicit_;
};

option::option(const std::string& key, const std::string& help)
    : d_(new d)
{
  d_->help_ = help;
  d_->narg_ = 1;
  d_->is_composing_ = false;
  d_->is_overwriting_ = false;
  d_->has_default_ = false;
  d_->has_implicit_ = false;

  add_key(key);
}

option::option(const option& o)
    : d_(new d(*o.d_))
{
}

option::option(option&& o)
  : d_(std::move(o.d_))
{
}

option& option::operator=(const option& o)
{
  if (this != &o)
    d_.reset(new d(*o.d_));
  return *this;
}

option::~option() {}

option& option::add_key(const std::string& k)
{
  if (!k.empty())
    d_->keys_.push_back(k);
  return *this;
}

const std::string& option::key() const
{
  return d_->keys_.empty() ? EMPTY : d_->keys_.front();
}

const std::vector<std::string>& option::keys() const
{
  return d_->keys_;
}

const std::string& option::help() const
{
  return d_->help_;
}

option& option::set_shortkey(const std::string& sk)
{
  d_->shortkeys_.clear();
  return add_shortkey(sk);
}

option& option::add_shortkey(const std::string& sk)
{
  if (!sk.empty())
    d_->shortkeys_.push_back(sk);
  return *this;
}

const std::string& option::shortkey() const
{
  return d_->shortkeys_.empty() ? EMPTY : d_->shortkeys_.front();
}

const std::vector<std::string>& option::shortkeys() const
{
  return d_->shortkeys_;
}

bool option::match(const std::string& key, bool use_shortkeys) const
{
  const std::vector<std::string>& k = use_shortkeys ? d_->shortkeys_ : d_->keys_;
  return std::find(k.begin(), k.end(), key) != k.end();
}

option& option::set_default_value(const std::string& d)
{
  d_->has_default_ = true;
  d_->default_ = d;
  return *this;
}

bool option::has_default_value() const
{
  return d_->has_default_;
}

const std::string& option::default_value() const
{
  return d_->default_;
}

option& option::set_implicit_value(const std::string& i)
{
  d_->has_implicit_ = true;
  d_->implicit_ = i;
  return *this;
}

bool option::has_implicit_value() const
{
  return d_->has_implicit_;
}

const std::string& option::implicit_value() const
{
  return d_->implicit_;
}

option& option::set_composing()

{
  d_->is_composing_ = true;
  d_->is_overwriting_ = false;
  return *this;
}

bool option::is_composing() const
{
  return d_->is_composing_;
}

option& option::set_overwriting()
{
  d_->is_composing_ = false;
  d_->is_overwriting_ = true;
  return *this;
}

bool option::is_overwriting() const
{
  return d_->is_overwriting_;
}

option& option::set_narg(size_t n)
{
  d_->narg_ = n;
  return *this;
}

size_t option::narg() const
{
  return d_->narg_;
}

} // namespace program_options
} // namespace miutil


mi_po_value_set.cc
/*
  mi-programoptions

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-programoptions.

  mi-programoptions is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-programoptions is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-programoptions; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_programoptions.h"

namespace miutil {
namespace program_options {

bool value_set::is_set(option_cx opt) const
{
  return opt && (values_.find(opt) != values_.end());
}

void value_set::put_implicit(option_cx opt)
{
  if (!opt)
    throw option_error("option is null");
  if (!opt->has_implicit_value())
    throw option_error("option '" + opt->key() + "' does not have an implicit value");
  if (opt->narg() != 1)
    throw option_error("option '" + opt->key() + "' expects != 1 values, cannot set to implicit value");
  put(opt, opt->implicit_value());
}

void value_set::put(option_cx opt, const string_v& values)
{
  if (!opt)
    throw option_error("option is null");
  if (opt->is_composing()) {
    if (values.size() != 1)
      throw option_error("option '" + opt->key() + "' is composing, cannot #values != 1");
  } else {
    if (is_set(opt) && !opt->is_overwriting())
      throw option_error("option '" + opt->key() + "' already set and not composing or overwriting");
  }

  string_v& v = values_[opt];
  if (opt->is_overwriting())
    v.clear();
  v.insert(v.end(), values.begin(), values.end());
}

const string_v* value_set::get(option_cx opt) const
{
  if (!opt)
    throw option_error("option is null");
  values_t::const_iterator it = values_.find(opt);
  if (it != values_.end())
    return &it->second;
  return nullptr;
}

const string_v& value_set::values(option_cx opt) const
{
  if (const string_v* values = get(opt))
    return *values;

  throw option_error("option '" + opt->key() + "' not set and without default");
}

const std::string& value_set::value(option_cx opt, size_t index) const
{
  if (const string_v* values = get(opt))
    return values->at(index);

  if (!opt)
    throw option_error("option is null");
  if (index == 0 && opt->has_default_value())
    return opt->default_value();

  throw option_error("option '" + opt->key() + "' not set and without default");
}

option_cx value_set::find(const std::string& key, bool use_shortkey) const
{
  for (const auto& o : values_) {
    const std::string& k = use_shortkey ? o.first->shortkey() : o.first->key();
    if (!k.empty() && k == key)
      return o.first;
  }
  return nullptr;
}

void value_set::add(const value_set& other)
{
  for (const auto& o : other.values_) {
    if (!values_.insert(o).second)
      throw option_error("option '" + o.first->key() + "' already set");
  }
}

} // namespace program_options
} // namespace miutil


mi_programoptions.cc
/*
  mi-programoptions

  Copyright (C) 2019 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-programoptions.

  mi-programoptions is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-programoptions is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-programoptions; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_programoptions.h"

#include <algorithm>
#include <fstream>
#include <sstream>

namespace {
const std::string EMPTY;
}

namespace miutil {
namespace program_options {

option::option(const std::string& key, const std::string& help)
    : help_(help)
    , narg_(1)
    , has_default_(false)
    , has_implicit_(false)
{
  add_key(key);
}

option::~option() {}

option& option::add_key(const std::string& k)
{
  if (!k.empty())
    keys_.push_back(k);
  return *this;
}

const std::string& option::key() const
{
  return keys_.empty() ? EMPTY : keys_.front();
}

option& option::set_shortkey(const std::string& sk)
{
  shortkeys_.clear();
  return add_shortkey(sk);
}

option& option::add_shortkey(const std::string& sk)
{
  if (!sk.empty())
    shortkeys_.push_back(sk);
  return *this;
}

const std::string& option::shortkey() const
{
  return shortkeys_.empty() ? EMPTY : shortkeys_.front();
}

bool option::match(const std::string& key, bool use_shortkeys) const
{
  const std::vector<std::string>& k = use_shortkeys ? shortkeys_ : keys_;
  return std::find(k.begin(), k.end(), key) != k.end();
}

bool value_set::is_set(option_cx opt) const
{
  return opt && (values_.find(opt) != values_.end());
}

void value_set::put_implicit(option_cx opt)
{
  if (!opt)
    throw option_error("option is null");
  if (!opt->has_implicit_value())
    throw option_error("option '" + opt->key() + "' does not have an implicit value");
  if (opt->narg() != 1)
    throw option_error("option '" + opt->key() + "' expects != 1 values, cannot set to implicit value");
  put(opt, opt->implicit_value());
}

void value_set::put(option_cx opt, const string_v& values)
{
  if (!opt)
    throw option_error("option is null");
  if (opt->is_composing()) {
    if (values.size() != 1)
      throw option_error("option '" + opt->key() + "' is composing, cannot #values != 1");
  } else {
    if (is_set(opt))
      throw option_error("option '" + opt->key() + "' already set and not composing");
  }

  string_v& v = values_[opt];
  v.insert(v.end(), values.begin(), values.end());
}

const string_v* value_set::get(option_cx opt) const
{
  if (!opt)
    throw option_error("option is null");
  values_t::const_iterator it = values_.find(opt);
  if (it != values_.end())
    return &it->second;
  return nullptr;
}

const string_v& value_set::values(option_cx opt) const
{
  if (const string_v* values = get(opt))
    return *values;

  throw option_error("option '" + opt->key() + "' not set and without default");
}

const std::string& value_set::value(option_cx opt, size_t index) const
{
  if (const string_v* values = get(opt))
    return values->at(index);

  if (!opt)
    throw option_error("option is null");
  if (index == 0 && opt->has_default_value())
    return opt->default_value();

  throw option_error("option '" + opt->key() + "' not set and without default");
}

option_cx value_set::find(const std::string& key, bool use_shortkey) const
{
  for (const auto& o : values_) {
    const std::string& k = use_shortkey ? o.first->shortkey() : o.first->key();
    if (!k.empty() && k == key)
      return o.first;
  }
  return nullptr;
}

void value_set::add(const value_set& other)
{
  for (const auto& o : other.values_) {
    if (!values_.insert(o).second)
      throw option_error("option '" + o.first->key() + "' already set");
  }
}

value_set parse_config_file(const std::string& filename, option_set& options)
{
  std::ifstream infile(filename);
  if (!infile)
    throw option_error("cannot read config file '" + filename + "'");

  try {
    return parse_config_file(infile, options);
  } catch (option_error& oe) {
    throw option_error("while reading '" + filename + ": " + oe.what());
  }
}

#define OPTION_CHARS "[a-zA-Z0-9._][a-zA-Z0-9._-]*"

value_set parse_config_file(std::istream& infile, option_set& options)
{
  value_set values;
  static const std::regex re_comment("\\s*#.*");
  static const std::regex re_section("\\[(" OPTION_CHARS ")\\]$");
  enum { GS_ALL, GS_KEY };
  static const std::regex re_value("\\s*(" OPTION_CHARS ")\\s*=\\s*(.*)?\\s*");
  enum { GV_ALL, GV_KEY, GV_VALUE };
  std::string line;
  std::string section;
  for (int lineno = 1; std::getline(infile, line); ++lineno) {
    if (line.empty())
      continue;
    std::smatch match;
    if (std::regex_match(line, match, re_comment)) {
      // ignore
    } else if (std::regex_match(line, match, re_section)) {
      section = match[GS_KEY].str() + ".";
    } else if (std::regex_match(line, match, re_value)) {
      std::string value = match[GV_VALUE].str();
      if (value.size() >= 2 && (value[0] == '\'' || value[0] == '"') && value[0] == value[value.size() - 1])
        value = value.substr(1, value.size() - 2);
      try {
        values.put(options.find_option(section + match[GV_KEY].str(), false), value);
      } catch (option_error& oe) {
        std::ostringstream msg;
        msg << "lineno " << line << ": " << oe.what();
        throw option_error(msg.str());
      }
    } else {
      std::ostringstream msg;
      msg << "bad line " << lineno << ": " << line;
      throw option_error(msg.str());
    }
  }
  if (!infile.eof() && infile.bad())
    throw option_error("error reading config");
  return values;
}

value_set parse_command_line(const std::vector<std::string>& argv, option_set& options, std::vector<std::string>& positional)
{
  value_set values;
  static const std::regex re_option("(-{1,2})(" OPTION_CHARS ")(=(.*))?");
  enum { GO_ALL, GO_DASH, GO_KEY, GO_EQ_VALUE, GO_VALUE };
  bool end_of_options_marker = false;
  const int argc = argv.size();
  for (int a = 0; a < argc; ++a) {
    const std::string arg(argv[a]);
    if (arg == "--") {
      end_of_options_marker = true;
      continue;
    }
    std::smatch match;
    if (!end_of_options_marker && std::regex_match(arg, match, re_option)) {
      option_cx opt = options.find_option(match[GO_KEY], match[GO_DASH].length() == 1);
      if (match[GO_VALUE].matched) {
        if (opt->narg() == 0) {
          throw option_error("arg for no-arg option '" + opt->key() + "'");
        } else if (!opt->is_composing() && opt->narg() != 1) {
          std::ostringstream msg;
          msg << "args for option '" + opt->key() + "', which expects exactly " << opt->narg() << " values, cannot be passed with '='";
          throw option_error(msg.str());
        }
        values.put(opt, match[GO_VALUE].str());
      } else if (opt->has_implicit_value()) {
        values.put_implicit(opt);
      } else if (opt->narg() == 0) {
        values.put(opt, std::string());
      } else if (opt->is_composing() && a + 1 < argc) {
        values.put(opt, argv[++a]);
      } else if (!opt->is_composing() && a + (int)opt->narg() < argc) {
        string_v args;
        args.reserve(opt->narg());
        const int alast = a + opt->narg();
        while (a < alast)
          args.push_back(argv[++a]);
        values.put(opt, args);
      } else {
        throw option_error("no arg for option '" + opt->key() + "'");
      }
    } else {
      positional.push_back(arg);
    }
  }
  return values;
}

value_set parse_command_line(int argc, char* argv[], option_set& options, std::vector<std::string>& positional)
{
  return parse_command_line(std::vector<std::string>(argv + 1, argv + argc), options, positional);
}

// ========================================================================

option_cx option_set::find_option(const std::string& key, bool use_shortkey)
{
  for (option_cx opt : options_) {
    if (opt->match(key, use_shortkey))
      return opt;
  }
  throw option_error("no such option '" + key + "'");
}

static void show_key(std::ostream& out, option_cx opt)
{
  bool first = true;
  const std::string sep = " / ";
  for (const auto& k : opt->keys()) {
    if (!first)
      out << sep;
    out << "--" << k;
    first = false;
  }
  for (const auto& sk : opt->shortkeys()) {
    if (!first)
      out << sep;
    out << "-" << sk;
    first = false;
  }
}

void option_set::dump(std::ostream& out, const value_set& values) const
{
  for (option_cx opt : options_) {
    if (values.is_set(opt)) {
      show_key(out, opt);
      out << std::endl;
      for (const std::string& v : *values.get(opt))
        out << "  => '" << v << "'" << std::endl;
    }
  }
}

void option_set::help(std::ostream& out) const
{
  for (option_cx opt : options_) {
    show_key(out, opt);
    out << ": " << opt->help();
    if (opt->has_default_value())
      out << " (default: " << opt->default_value() << ")";
    out << std::endl;
  }
}

positional_args_consumer& positional_args_consumer::operator>>(const option& opt)
{
  if (opt.is_composing()) {
    values_.put(&opt, *pbegin_++);
  } else if (!values_.is_set(opt)) {
    if (std::distance(begin(), end()) >= (int)opt.narg()) {
      const string_v::const_iterator opt_end = begin() + opt.narg();
      values_.put(&opt, string_v(begin(), opt_end));
      pbegin_ = opt_end;
    } else {
      throw option_error("positional arg error for option '" + opt.key() + "'");
    }
  }
  return *this;
}

void positional_args_consumer::dump(std::ostream& out) const
{
  out << "remaining positional arguments: " << std::endl;
  for (string_v::const_iterator p = pbegin_; p != positional_.end(); ++p)
    out << " -> '" << *p << "'" << std::endl;
}

} // namespace program_options
} // namespace miutil
